core/store/src/flat/manager.rs

```diff
@@ -93,6 +96,107 @@ impl FlatStorageManager {
+                        debug!(
+                            target: "chain",
+                            ?new_flat_head,
+                            ?err,
+                            ?shard_uid,
+                            block_hash = ?block.header().hash(),
+                            "Cannot update flat head");
+                    }
+                    _ => {
+                        // All other errors are unexpected, so we panic.
+                        panic!("Cannot update flat head of shard {shard_uid:?} to {new_flat_head:?}: {err:?}");
+                    }
+                }
+            });
+        } else {
+            // TODO (#8250): come up with correct assertion. Currently it doesn't work because runtime may be
+            // implemented by KeyValueRuntime which doesn't support flat storage, and flat storage background
+            // creation may happen.
+            // debug_assert!(false, "Flat storage state for shard {shard_id} does not exist and its creation was not initiated");
+        }
+        Ok(())
+    }
+
+    pub fn save_flat_state_changes(
+        &self,
+        block_hash: CryptoHash,
+        prev_hash: CryptoHash,
+        height: BlockHeight,
+        shard_uid: ShardUId,
+        state_changes: &Vec<RawStateChangesWithTrieKey>,
+    ) -> Result<StoreUpdate, StorageError> {
+        let prev_block_with_changes = if state_changes.is_empty() {
+            // The current block has no flat state changes.
+            // Find the last block with flat state changes by looking it up in
+            // the prev block.
+            store_helper::get_prev_block_with_changes(
+                &self.0.store,
+                shard_uid,
+                block_hash,
+                prev_hash,
+            )
+            .map_err(|e| StorageError::from(e))?
+        } else {
+            // The current block has flat state changes.
+            None
+        };
+
+        let delta = FlatStateDelta {
+            changes: FlatStateChanges::from_state_changes(state_changes),
+            metadata: FlatStateDeltaMetadata {
+                block: BlockInfo { hash: block_hash, height, prev_hash },
+                prev_block_with_changes,
+            },
+        };
+
+        let store_update = if let Some(flat_storage) = self.get_flat_storage_for_shard(shard_uid) {
+            // If flat storage exists, we add a block to it.
+            flat_storage.add_delta(delta).map_err(|e| StorageError::from(e))?
+        } else {
+            let shard_id = shard_uid.shard_id();
+            // Otherwise, save delta to disk so it will be used for flat storage creation later.
+            debug!(target: "chain", %shard_id, "Add delta for flat storage creation");
+            let mut store_update = self.0.store.store_update();
+            store_helper::set_delta(&mut store_update, shard_uid, &delta);
+            store_update
+        };
+
+        Ok(store_update)
```

Comment:
Note: This part is also slightly changed from original implementation. Here we are returning the store_update and it's the responsibility of the caller to apply/commit changes (merge with chain_store_update).